### ISSUE TITLE
Declarative modules next steps

### ISSUE THREAD

**Issue by Tpt** [PyO3/pyo3#3900]
declarative modules with the `#[pymodule] mod X {}` syntax is now merged behind the `experimental-declarative-modules` syntax (#3815).

The next steps are:

Before v0.21:
- [x] document the new feature in the guide
- [x] allow to export `#[pyclass]`, `#[pyfunction]` and `#[pymodule]` that are inlined in the modules (in progress in #3902)

Before v0.22:
- [ ] properly register submodules in `sys.modules` automatically i.e. fix #759 by default
- [ ] allow `#[pymodule_init]` function to return `()` or `PyResult<()>`
- [x] adding a class to a module via the `PyAddToModule` trait is currently relying on `PyTypeInfo::type_object_bound`. This function panics when module initialization fail. It would be better to throw an exception instead like in `PyModuleMethods::add_class` (in progress in #3919)
- [ ] investigate relationship between this and "two-phase initialization" e.g. #2245. In particular we should look at `#[pymodule_data]` as a way to create a per-module data mechanism (might be postpone to later)
- [ ] allow `#[pymodule_export] const FOO = ...;` to add constants to the modules
- [ ] investigate automatically filling the `module` parameter of `#[pyclass]` usages inside of decarative modules. #4213
- [x] remove the  `experimental-declarative-modules` feature flag to allow using declarative modules by default
- [x] update examples, README and guide

Before v0.23 (or later?):
- [x] Deprecate `#[pyfn]`
- [ ] Deprecate "function-based" modules

**Comment by davidhewitt** [comment]
Possibly worth us remembering the discussion in https://github.com/PyO3/pyo3/pull/3919#discussion_r1515103061 to decide what is allowed to export before we declare stable and remove the feature gate.

**Comment by nickdrozd** [comment]
Some user feedback: I just upgraded to `0.21.0-beta.0`, and the declarative module feature is very nice. It allows for getting rid of a bunch of worthless junk like `m.add_function(wrap_pyfunction!(some_func, m)?)?;`. Maybe I hate repetitive boilerplate more than most, but this was a big relief for me.

Of course you have all had a lot of discussion about this, and I don't understand most of the details. But from my perspective, the best course would be to bless declarative modules as the standard way of doing things and throw all that old `m.add_function` stuff in the garbage ASAP.

It's not just that it's a nice feature. It's also that it might end up being on the migration path for a lot of users anyway. Moving from 20 to 21 raised several compiler warnings, and the only way I could figure out to silence all of them was to use a declarative module.

**Comment by davidhewitt** [comment]
Hey @nickdrozd, thanks for the feedback! I'm pleased you like these, I agree that imperative modules should be deprecated and replaced with these modules as soon as possible. This feature is probably the thing I've wanted most for PyO3 for the longest, except for the `Bound` API we're finally shipping in 0.21.

We _could_ just call these stable in 0.21, but there are a couple of corner cases of unresolved design that I think could break people so I'd like to get these at least understood before we tell people it's stable. From the guide that we'll ship in 0.21:

> Some changes are planned to this feature before stabilization, like automatically filling submodules into sys.modules to allow easier imports (see [issue #759](https://github.com/PyO3/pyo3/issues/759)) and filling the module argument of inlined #[pyclass] automatically with the proper module name. Macro names might also change.

**Comment by davidhewitt** [comment]
Just to note against the alternative of 0.21 delaying until these design points are ready: so much is changing in 0.21 already that I'm quite keen to see the release finalised so that users can handle these. There's a lot of stuff beginning to pile up that I've asked to wait until 0.22 already, it would be good to unblock the development flow.

If it turns out that lots of users need to update to declarative modules to solve compile issues, then I think probably we got something wrong with the design of the Bound API migration 😢. Hopefully the additional deprecation warnings we're shipping in 0.21 final compared to 0.21 beta will make the difference.

**Comment by nickdrozd** [comment]
> If it turns out that lots of users need to update to declarative modules to solve compile issues, then I think probably we got something wrong with the design of the Bound API migration 😢. Hopefully the additional deprecation warnings we're shipping in 0.21 final compared to 0.21 beta will make the difference.

I had a setup like this:

```rust
#[pymodule]
fn rust_stuff(py: Python, m: &PyModule) -> PyResult<()> {
    // ...
}
```

After switching to `0.21.0-beta.0`, I got this compiler warning:

```
warning: use of deprecated method `pyo3::methods::Extractor::<T>::extract_gil_ref`: use `&Bound<'_, T>` instead for this function argument
  --> src/lib.rs:30:27
   |
30 | fn rust_stuff(py: Python, m: &PyModule) -> PyResult<()> {
   |                           ^^^^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default
```

I don't know anything about `Bound` or `extract_gil_ref`, so the warning left me baffled. It seemed easier just to switch to the declarative style.

(I am using PyO3 to rewrite a Python project in Rust using a bottom-up leaf-first approach. So Python calls into Rust, but the Rust code never interacts with Python except through `pyfunction`, `pyclass`, etc).

**Comment by davidhewitt** [comment]
Do you think it would have helped if these deprecation messages link you to the PyO3 migration guide?

**Comment by nickdrozd** [comment]
> Do you think it would have helped if these deprecation messages link you to the PyO3 migration guide?

I check the migration guide, but it doesn't say anything about how to update `m: &PyModule`.

**Comment by alecandido** [comment]
I'm a bit puzzled by the submodule. In the example on the docs is left empty

but as soon as I try to add something, it fails:
```rs
#[pymodule]
mod submodule {
    #[pyfunction]     ■■ consider importing one of these items: `use crate::pyfunction;    `, `use pyo3::pyfunction;    `
    pub fn triple(x: usize) -> usize {     ■ failed to resolve: function `triple` is not a crate or module  function `triple`
        x * 3
    }
}
```
```sh
error: cannot find attribute `pyfunction` in this scope
  --> src/lib.rs:24:11
   |
24 |         #[pyfunction]
   |           ^^^^^^^^^^
   |
help: consider importing one of these items
   |
24 +         use crate::pyfunction;
   |
24 +         use pyo3::pyfunction;
   |
                                                                                                                                         
error[E0433]: failed to resolve: function `triple` is not a crate or module
  --> src/lib.rs:25:16
   |
25 |         pub fn triple(x: usize) -> usize {
   |                ^^^^^^ function `triple` is not a crate or module
```

and without the `#[pyfunction]` it simply doesn't do anything...
(i.e. it's not accessible from the Python `module.submodule.triple`)

How is it supposed to be used?

**Comment by alecandido** [comment]
Moreover, something weird happens with classes. This is working as expected:
```rs
use pyo3::prelude::*;

mod othermod {
    use pyo3::prelude::*;

    #[pyfunction]
    pub fn triple(x: usize) -> usize {
        x * 3
    }
}

#[pymodule]
mod myext {
    #[pymodule_export]
    use super::othermod::triple;
}
```
but this is not:
```rs
use pyo3::prelude::*;
                                                                                           
mod othermod {
    use pyo3::prelude::*;
                                                                                           
    #[pyclass]    // ■ private associated constant defined here
    pub struct Unit;
}
                                                                                           
#[pymodule]    // ■■ associated constant `_PYO3_DEF` is private  private associated constant
mod myext {               
    #[pymodule_export]
    use super::othermod::Unit;
}
```
```sh
error[E0624]: associated constant `_PYO3_DEF` is private
  --> src/lib.rs:10:1
   |
6  |     #[pyclass]
   |     ---------- private associated constant defined here
...
10 | #[pymodule]
   | ^^^^^^^^^^^ private associated constant
   |
   = note: this error originates in the attribute macro `pymodule` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by alecandido** [comment]
Oh, thanks. I actually searched for `_PYO3_DEF` in the repo, but I forgetting the leading `_` (and it didn't show the issue...)

https://github.com/search?q=repo%3APyO3%2Fpyo3+PYO3_DEF&type=issues
https://github.com/search?q=repo%3APyO3%2Fpyo3+_PYO3_DEF&type=issues

Sorry...

**Comment by wyfo** [comment]
Just a remarks exception created with `create_exception!` are not mentioned in this issue. The only solution I see for the moment is to declare exception outside the module and use `#[pymodule_export]`.
I don't have any idea on how we could improve ergonomics for now.

**Comment by Tpt** [comment]
> Just a remark: exception created with `create_exception!` are not mentioned in this issue. The only solution I see for the moment is to declare exception outside the module and use `#[pymodule_export]`.

With the current code, yes! I would tend to think the long term solution is #295 but it seems quite a bit of work. An other slighty hacky approach might be to make `#[pymodule]` macro automatically expose all `create_exception!` calls inside of the module.

**Comment by Stargateur** [comment]
Documentation on the module are ignored:

```rust
#[pymodule]
/// This is ignored
pub mod foo {
}
```

**Comment by Tpt** [comment]
@Stargateur Thank you for reporting this! It is weird, we have [a test covering exactly this](https://github.com/PyO3/pyo3/blob/37a5f6a94e9dab31575b016a4295fb94322b9aba/tests/test_declarative_module.rs#L131). May you share a bit more how you load the module and try to extract the documentation?

**Comment by Stargateur** [comment]
my MCE is as follow:

```rust
//! Foo
#![warn(missing_docs)]
use pyo3::pymodule;


#[pymodule]
/// This doesn't remove rustc warning about missing docs
pub mod foo {
}
```

```toml
[package]
name = "tmp"
version = "0.1.0"
edition = "2021"

[dependencies.pyo3]
version = "0.21.2"
features = [
    "experimental-declarative-modules",
]
```

```none
warning: missing documentation for a module
 --> src\lib.rs:6:1
  |
6 | #[pymodule]
  | ^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src\lib.rs:2:9
  |
2 | #![warn(missing_docs)]
  |         ^^^^^^^^^^^^
  = note: this warning originates in the attribute macro `pymodule` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `tmp` (lib) generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
```

And `cargo doc` doesn't show the documentation of module `foo` either.

**Comment by Tpt** [comment]
@Stargateur Thank you! So, I guess the `#[pymodule]` macro is not properly reemiting the doc comment on the module. I am going to investigate that.

**Comment by SichangHe** [comment]
(Not sure if relevant) Just tried to feature gate the module macros, but met trouble since `pymodule_export` is not a real macro, just an attribute:

```rust
#[cfg(feature = "pyo3")]
use pyo3::prelude::*;

#[cfg_attr(feature = "pyo3", pymodule)]
pub mod py_const {
    #[cfg_attr(feature = "pyo3", pymodule_export)]
    pub const MY_CONST: () = ();
}
```

cannot find attribute `pymodule_export` in this scope

**Comment by SichangHe** [comment]
<details><summary>*Edit*: This is wrong. This it equivalent the `[pymodule_export]` not being there. What needs to be done is another `pymodule` macro…</summary>

Okay, seems it's a small problem. All I needed was importing a dummy attribute procedure macro and renaming it to `pymodule_export`:

```rust
#[cfg(feature = "pyo3")]
use pyo3::prelude::*;

#[cfg_attr(feature = "pyo3", pymodule)]
pub mod py_const {
    #[cfg(feature = "pyo3")]
    use pass_through_macro::pass_through as pymodule_export;
    #[cfg_attr(feature = "pyo3", pymodule_export)]
    pub const MY_CONST: () = ();
}
```

<details><summary>The implementation of <code>pass_through</code> in case anyone is interested.</summary>

```rust
// lib.rs
use proc_macro::TokenStream;

/// Dummy attribute macro that passes through the input out.
#[proc_macro_attribute]
pub fn pass_through(_attr_tokens: TokenStream, input_tokens: TokenStream) -> TokenStream {
    input_tokens
}
```

```toml
# Cargo.toml
[package]
name = "pass_through_macro"

[lib]
proc-macro = true
```

</details>
</details>

**Comment by SichangHe** [comment]
*Edit:* this only works with the feature enabled…

Dumb me. Simply this would work. Sorry for the noise.

```rust
#[cfg(feature = "pyo3")]
use pyo3::prelude::*;

#[cfg_attr(feature = "pyo3", pymodule)]
pub mod py_const {
    #[pymodule_export]
    pub const MY_CONST: () = ();
}
```

**Comment by SichangHe** [comment]
Okay, still a small problem. All I needed was, when `pyo3` is *not active*, importing a dummy attribute procedure macro, renaming it to `pymodule_export`, and using it:

```rust
#[cfg(feature = "pyo3")]
use pyo3::prelude::*;

#[cfg_attr(feature = "pyo3", pymodule)]
pub mod py_const {
    #[cfg(not(feature = "pyo3"))]
    use macros::pass_through as pymodule_export;
    #[pymodule_export]
    pub const MY_CONST: () = ();
}
```

<details><summary>The implementation of <code>pass_through</code> and the <code>Cargo.toml</code>s in case anyone is interested.</summary>

```rust
// macros/src/lib.rs
use proc_macro::TokenStream;

/// Dummy attribute macro that passes through the input out.
#[proc_macro_attribute]
pub fn pass_through(_attr_tokens: TokenStream, input_tokens: TokenStream) -> TokenStream {
    input_tokens
}
```

```toml
# macros/Cargo.toml
[package]
name = "macros"

[lib]
proc-macro = true
```

```toml
# my_crate/Cargo.toml
[package]
name = "my_crate"

[lib]
name = "my_crate"
crate-type = ["lib", "cdylib"]

[dependencies]
macros.workspace = true

pyo3 = { workspace = true, optional = true }

[features]
default = ["pyo3"]
pyo3 = ["dep:pyo3"]
```

</details>

**Comment by Tpt** [comment]
@SichangHe Thank you for reporting this. I have opened a specific issue for that: #5183

**Comment by MarkusSintonen** [comment]
Is it possible to get the pyclass module name set correctly in declarative modules? Now everything being under `builtin` confuses eg document generators (pdoc / mkdocs). Also it is pretty awkward to use `#[pyclass(module="xxx")]` as you need to define the modules 2x.

**Comment by Icxolu** [comment]
I believe if the pyclass is defined inline in the declarative module, we already do that, but if you export the class via
`#[pymodule_export]` this is not possible, and you have to do it manually.

**Comment by elbaro** [comment]
The submodule name is missing the parent modules.

```rs
#[pymodule]
mod mylib {
    #[pymodule]
    mod mymodule {
    }
}
```

```sh
>>> import collections.abc
>>> collections.abc.__name__
'collections.abc'

>>> import mylib.mymodule
>>> mylib.mymodule
'mymodule'  # should be 'mylib.mymodule'
```

And this can't be customized with `#[pymodule(name = "mylib.mymodule")]`: `expected a single identifier in double quotes`

**Comment by davidhewitt** [comment]
Looks like #5375 will fix this (thanks @altendky).

**Comment by stijndcl** [comment]
Is there a way to properly (albeit manually) configure nested module paths when splitting declarative modules across multiple files? This seems like a desirable pattern for larger projects where doing everything in one file is not feasible anymore.

<details>
<summary>Code sample</summary>

```rs
/// File 1 (root.rs)
#[pymodule]
pub(crate) mod root {
    #[pymodule_export]
    use crate::config::config;
}
```

```rs
/// File 2 (config.rs)

#[pymodule(submodule)]
pub(crate) mod config {
    #[pymodule_export]
    use crate::subconfig::subconfig;
}

/// File 3 (subconfig.rs)

#[pymodule(submodule)]
pub(crate) mod subconfig {}
```
</details>

In this example, everything inside of `config` here is defined as `config.X` and `subconfig.X`, while in practice it should be `root.config.X` and `root.config.subconfig.X`, respectively. As mentioned [a few comments above](https://github.com/PyO3/pyo3/issues/3900#issuecomment-3362121331) the module naming is passed down automatically for pyclasses and pyfunctions that are added inline. In this case it's still just wrong because the containing _modules themselves_ are not defined inline in their parent module, which makes sense I suppose.

I understand this macro not being able to automatically figure out that it is nested inside of `root` (or the other way around - the `root` macro not being able to pass that info downstream to `config`), so I assume some manual effort will be involved, but as far as I can tell you can't really do that right now:

<details>
<summary>Code sample</summary>

```rs
#[pymodule(submodule, name = "root.config")]
pub(crate) mod config {}
```
</details>

Currently this errors with `expected a single identifier in double quotes`. I understand from the docs that you _can_ do it for function-style modules using the `pyo3(module = "x.y")` macro, but I don't see a similar option for declarative modules atm, and it seems like declarative is the way forward. I tried to read the expanded macro and it looks like the module name is stored as a string so this might be possible? If the `name` really has to be a valid identifier, perhaps an additional attribute `module = "any.string"`, or `parent = "any.string"` (which auto-appends `.{name} to it`) could help instead?

Another option could be to allow referencing the parent module somehow to make it slightly less tedious/error-prone, eg

<details>
<summary>Code sample</summary>

```rs
// Can this read <parent>::_PYO3_NAME or similar to construct its own _PYO3_NAME?
#[pymodule(parent = crate::root::root)] // `parent` possibly implies `submodule`?
pub(crate) mod config {}

#[pymodule(parent = crate::config::config)] // can do this for nested submodules to pull the full path from root
pub(crate) mod subconfig {}
```
</details>

but I have no idea if that is even possible.

Sidenote: the top of this ticket lists `properly register submodules in sys.modules automatically`. As far as I can tell you need the full path for that anyways so I feel it's directly related to/blocked by this.

Thanks and sorry for the rant 😅

**Comment by Tpt** [comment]
@stijndcl Thank you! This is a great point. I like 
```rust
#[pymodule(parent = crate::config::config)] // can do this for nested submodules to pull the full path from root
pub(crate) mod subconfig {}
```
because it makes things very explicit and breaks the build or migrates transparently if the module name changes. I did not thought about it, if it actually works it can be a great option.

**Comment by stijndcl** [comment]
I don't know if it will work, hence the

> but I have no idea if that is even possible.

Macros aren't my forte just yet, I was just spitballing ideas for things that might help. Perhaps this exact syntax is not possible, but something like `crate::config::PYO3_NAME` is instead. I'll see if I can play around with it the coming days to make something reasonable work.

**Comment by Tpt** [comment]
> I'll see if I can play around with it the coming days to make something reasonable work.

Amazing! Ping me if you need help or reviews!

**Comment by stijndcl** [comment]
@Tpt Hi I've done some digging around and I've concluded two things

- Directly referring to the parent module/its name is unfortunately not possible as far as I can tell
- PyO3 already supports what I'm trying to do, but it's not documented

When looking at the macro I ran across the `module = "string"` arg which seems to be exactly what I was asking for in my comment above, I just mistakenly guessed `name` which serves an entirely different purpose. Looking back at it, it was quite obvious because the old function syntax supported `pyo3(module = "")`. Unfortunately this is a hardcoded string that people will have to keep in sync themselves, but realistically I doubt this will change often anyways.

Apologies. Wouldn't hurt to add this to the docs for modules though I feel, so perhaps I can contribute that instead?

---

With that being said, it _should_ now be possible to auto-generate the (hacky) `pymodule_init` step that registers your module to `sys.modules`:

```rs
    #[pymodule_init]
    fn init(m: &Bound<'_, PyModule>) -> PyResult<()> {
        Python::attach(|py| {
            py.import("sys")?
                .getattr("modules")?
                .set_item("root.submodule", m)
        })
    }
```

The path `root.submodule` is exactly the same as `{module}.{name}` which seems to be `full_name` inside of the macro, so that info is already available.

**Comment by Tpt** [comment]
@stijndcl Thank you for investigating. yes, updating the doc would be amazing!
